src/devices/bus/a2bus/vistaa800.cpp

```diff
@@ -0,0 +1,327 @@
+
+const tiny_rom_entry *a2bus_vistaa800_device::device_rom_region() const
+{
+	return ROM_NAME(vistaa800);
+}
+
+void a2bus_vistaa800_device::device_add_mconfig(machine_config &config)
+{
+	FD1797(config, m_fdc, 2000000);
+	FLOPPY_CONNECTOR(config, "fdc:0", vistaa800_floppies, "8dsdd", floppy_image_device::default_mfm_floppy_formats).enable_sound(true);
+	FLOPPY_CONNECTOR(config, "fdc:1", vistaa800_floppies, "8dsdd", floppy_image_device::default_mfm_floppy_formats).enable_sound(true);
+	FLOPPY_CONNECTOR(config, "fdc:2", vistaa800_floppies, "8dsdd", floppy_image_device::default_mfm_floppy_formats).enable_sound(true);
+	FLOPPY_CONNECTOR(config, "fdc:3", vistaa800_floppies, "8dsdd", floppy_image_device::default_mfm_floppy_formats).enable_sound(true);
+
+	m_fdc->intrq_wr_callback().set(FUNC(a2bus_vistaa800_device::fdc_intrq_w));
+	m_fdc->drq_wr_callback().set(FUNC(a2bus_vistaa800_device::fdc_drq_w));
+	m_fdc->sso_wr_callback().set(FUNC(a2bus_vistaa800_device::fdc_sso_w));
+}
+
+void a2bus_vistaa800_device::device_start()
+{
+	save_item(NAME(m_dmaaddr));
+	save_item(NAME(m_dmaenable_read));
+	save_item(NAME(m_dmaenable_write));
+	save_item(NAME(m_density));
+	save_item(NAME(m_side));
+	save_item(NAME(m_sso));
+}
+
+void a2bus_vistaa800_device::device_reset()
+{
+	m_dmaaddr = 0;
+	m_dmaenable_read = false;
+	m_dmaenable_write = false;
+	m_density = 0;
+	m_side = 0;
+	m_sso = 0;
+}
+
+//----------------------------------------------
+//  device_a2bus_card_interface implementation
+//----------------------------------------------
+
+uint8_t a2bus_vistaa800_device::read_c0nx(uint8_t offset)
+{
+	uint8_t result = 0;
+
+	switch (offset)
+	{
+		case 0:
+		case 1:
+		case 2:
+		case 3:
+		case 4:
+		case 5:
+		case 6:
+		case 7:
+			result = m_fdc->fd1797_device::read(offset & 0x03);
+			break;
+			
+		case 0xa:
+			m_dmaenable_read = true;
+			break;
+
+		case 0xb:
+			m_dmaenable_write = true;
+			break;
+
+		case 0xc:
+			m_dmaenable_read = false;
+			m_dmaenable_write = false;
+			break;
+
+		case 0xf:
+			if (m_dmaenable_read || m_dmaenable_write)
+			{
+				result = result | 0x80;
+			}
+
+			result = result | (m_side << 5); // Todo: check this
+			break;
+
+		default:
+			logerror("Read c0n%x (%s)\n", offset, machine().describe_context());
+			break;
+	}
+
+	return result;
+}
+
+void a2bus_vistaa800_device::write_c0nx(uint8_t offset, uint8_t data)
+{
+	floppy_image_device *floppy = nullptr;
+
+	switch (offset)
+	{
+		case 0:
+		case 1:
+		case 2:
+		case 3:
+		case 4:
+		case 5:
+		case 6:
+		case 7:
+			m_fdc->fd1797_device::write(offset & 0x03, data);
+			break;
+			
+		case 8:
+			m_dmaaddr = (m_dmaaddr & 0xff00) + data;
+			break;
+
+		case 9:
+			m_dmaaddr = (m_dmaaddr & 0x00ff) + (data << 8);
+			break;
+
+		case 0xa:
+			m_dmaenable_read = true;
+			break;
+
+		case 0xb:
+			m_dmaenable_write = true;
+			break;
+
+		case 0xc:
+			m_dmaenable_read = false;
+			m_dmaenable_write = false;
+			break;
```

Comment:
You forgot to guard against debugger accesses here.

Reply:
Could you explain this a bit more, is the guard for the debugger access needed for reads or writes?

Reply:
It's typically only required for reads, because memory views can generated multiple unexpected reads and cause unwanted state change. On the other hand, writes from the debugger are often intended to trigger such state changes, and won't happen without an explicit user action.

Reply:
See here for implementation for this device: https://github.com/mamedev/mame/blob/62c0904ba46c497bc45605709a02963b582e14ea/src/devices/bus/a2bus/vistaa800.cpp#L168

Read this section of the documentation for an explanation from the opposite direction (i.e. from a debugger user’s point of view): https://docs.mamedev.org/debugger/#memory-accesses

In simple terms, debugger memory and disassembly windows generate read accesses in order to display the contents of address spaces as seen by the emulated system.  If read accesses have side effects, you need to guard against debugger accesses or simply having debugger windows open will change the state of the emulated system.